File: src/main/go/leetcode/lists/hot100/LeetCode437.go
package hot100

func pathSum(root *TreeNode, targetSum int) int {
	if root == nil {
		return 0
	}
	var dfs func(*TreeNode, int) int
	dfs = func(root *TreeNode, targetSum int) int {
		var count int
		if root == nil {
			return 0
		}
		if root.Val == targetSum {
			count++
		}
		count += dfs(root.Left, targetSum-root.Val)
		count += dfs(root.Right, targetSum-root.Val)
		return count
	}
	return dfs(root, targetSum) + pathSum(root.Left, targetSum) + pathSum(root.Right, targetSum)
}

File: src/main/java/leetcode/lists/hot100/LeetCode437.java
package leetcode.lists.hot100;


import common.TreeNode;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/10 01:29
 */
public class LeetCode437 {

    //https://www.youtube.com/watch?v=1S7j67cREZ8
    public int pathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return 0;
        }
        return rootSum(root, targetSum) +// 首先写root节点
                // 扩展到其他节点
                pathSum(root.left, targetSum) +
                pathSum(root.right, targetSum);
    }

    public int rootSum(TreeNode root, long targetSum) {
        int ret = 0;

        if (root == null) {
            return 0;
        }
        if (root.val == targetSum) {
            ret++;
        }

        ret += rootSum(root.left, targetSum - root.val);
        ret += rootSum(root.right, targetSum - root.val);
        return ret;
    }
}
